### DISCUSSION TITLE
Really struggling with this crate right now and need help

### DISCUSSION THREAD

**Discussion by askor** [idanarye/bevy-tnua#41] [Q&A] [answered]
Hello!

I was a previous user of Bevy Mod Wanderlust, but the dev has abandoned it for this crate it seems.

It's a bit embarrassing, but I've been trying for a few days now and I just can't get this crate to work. The example code is very hard to read with #[cfg()] all over the place and everything is split into many packages. There is even a example crate used in the example file, and the example crate is also hard to read. Is there some minimal example?

Right now I have this:

```
pub struct PlayerPlugin;

impl Plugin for PlayerPlugin {
    fn build(&self, app: &mut App) {
        app
            // Plugins
            .add_plugins(TnuaXpbd3dPlugin)
            .add_plugins(TnuaControllerPlugin)
            .add_plugins(TnuaCrouchEnforcerPlugin)
            // Systems
            .add_systems(Startup, setup_player)
            ;
    }
}

pub(crate) fn setup_player( mut commands: Commands) {
    commands.spawn((
        Camera3dBundle {
            transform: Transform::from_xyz(1., 40., 0.).looking_at(Vec3::ZERO, Vec3::Y),
            ..default()
        },
        RigidBody::Dynamic,
        Collider::capsule(1.0, 0.5),
        TnuaControllerBundle::default(),
    ));
}
```

In the example, there is also a component CharacterMotionConfigForPlatformerExample, but I don't know how to get the example crate and I can't understand what this component does. There is a comment, but I honestly don't understand what it is trying to say:

    // This bundle container `TnuaController` - the main interface of Tnua with the user code - as
    // well as the main components used as API between the main plugin and the physics backend
    // integration. These components (and the IO bundle, in case of backends that need one like
    // Rapier) are the only mandatory Tnua components - but this example will also add some
    // components used for more advanced features.

What happens when I run my game is that the camera just falls and there is no floating and no locking of axes, so it just falls over. Maybe I'm stupid or bad Rust programmer, but I'm not sure why this crate is so complex. With Wanderlust you just add a plugin and a component to you player entity. I hate to say it, but I regret the time I spent doing this in stead of just programming my own character from scratch. I think the point of using a crate in stead of doing it yourself is that you can save time to just get started, but eventually most people will probably have to program their own character controller anyways.

Sorry this is very negative, and I respect the work into this crate. Looks like a good controller, I just think it needs simpler examples, maybe without support for xpbd and rapier in the same file, and less stuff to get started. If you can help with a minimal example, I would be grateful, and I will maybe use bevy_fpc in the mean time.

**Comment by idanarye** [comment]
I get what you are saying. I think maybe I was too focused on having the examples demonstrate everything Tnua has to offer, and by doing that sacrificing the readability. I was warned about it, but I guess I just dismissed the warnings... I'll try making a basic quick-start example that only shows how to do simple walk and jump.

`CharacterMotionConfigForPlatformerExample` is not something that you need to use in your own game. It's the configuration I use in the example, and contains lots of things that in real game could just be hard-coded. You also don't need `TnuaCrouchEnforcerPlugin` unless you use Tnua's crouching feature. The other plugins you do need.

The thing you are missing here is a character control system. [The full example is quite big](https://github.com/idanarye/bevy-tnua/blob/main/examples/src/character_control_systems/platformer_control_systems.rs) (it does try to demonstrate everything), but at the minimum it needs, every frame, to get the `TnuaController` component of the character and call:

```rust
controller.basis(TnuaBuiltinWalk {
    desired_velocity: some_vec3_with_the_velocity_you_want_to_go_in,
    // desired_forward: you only have to set this if you want Tnua to rotate the character on the Y axis
    float_height: the_height_the_character_should_float_at,
    ..Default::default()
});
```

**Reply by idanarye** [reply, marked as answer]
I've renamed the "examples" to "demos" (which better describe what they do) and added a smaller example instead that can serve as a proper quickstart: https://github.com/idanarye/bevy-tnua/blob/main/examples/example.rs

@askor - please check it out and tell me if it helps you understand how to use this plugin.

**Reply by askor** [reply]
Absolutely brilliant, this is what I needed! Thank you :) Much easier to follow and clearly demonstrates how to get started. Got it working in less than a minute.

If I can suggest it, maybe add some pre-made first and third person controllers, with camera and keys pre-configured. It's very nice when making a new project to just add a crate and start walking in first person without fiddling too much with copying code. Then if people need to tune stuff, they can look at examples.

Either way, this is much easier for small brained people like me. Thanks!

**Reply by askor** [reply]
Oh, and btw, maybe add something to readme that says you need both bevy-tnua crate and bevy-tnua-xxxx crate to really spoon feed people like me

**Reply by idanarye** [reply]
It's already there, at the beginning of the README:

![image](https://github.com/idanarye/bevy-tnua/assets/1149255/f520f1f5-d863-4049-b1c6-5bb50bc64b4f)

Is this not enough?

**Reply by askor** [reply]
My first thought was that I _only_ needed bevy-tnua-xpbd3d when I read that you see.

**Reply by idanarye** [reply]
Okay. I'll add a clarification.

**Reply by idanarye** [reply]
Done
